File: 1_Two_Sum/main.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

vector<int> twoSum(vector<int>& nums, int target) {
    unordered_map<int, int> tmp;
    for (int i = 0; i < nums.size(); ++i) {
        auto it = tmp.find(target - nums[i]);
        if (it != tmp.end()) {
            return {it->second, i};
        }
        tmp[nums[i]] = i;
    }
    return {};
}

int main() {
    vector<int> v = {3, 3};
    int target = 6;
    auto res = twoSum(v, target);
    for (auto e : res) {
        cout << e << endl;
    }
    return 0;
}

File: 1_Two_Sum/golang/main.go
package golang

func twoSum(nums []int, target int) []int {
	result := make([]int, 0, 2)

	m := make(map[int]int, len(nums))

	for i, v := range nums {
		if 2*v == target {
			result = append(result, i)
			if len(result) == 2 {
				return result
			}
		} else {
			m[v] = i
		}
	}

	for k, v := range m {
		need := target - k

		if j, ok := m[need]; ok {
			return []int{v, j}
		}
	}

	return nil
}

func twoSumV2(nums []int, target int) []int {
	m := make(map[int]int, len(nums))

	for i, num := range nums {
		if j, ok := m[target - num]; ok {
			return []int{i, j}
		}

		m[num] = i
	}

	return nil
}